dimod/include/dimod/constrained_quadratic_model.h

```diff
@@ -0,0 +1,493 @@
+
+            assert(vartype != Vartype::SPIN || lb == -1);
+            assert(vartype != Vartype::SPIN || ub == +1);
+        }
+
+        explicit varinfo_type(Vartype vartype)
+                : vartype(vartype),
+                  lb(vartype_info<bias_type>::default_min(vartype)),
+                  ub(vartype_info<bias_type>::default_max(vartype)) {}
+    };
+
+    std::vector<varinfo_type> varinfo_;
+};
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>::ConstrainedQuadraticModel()
+        : objective(this), constraints_(), varinfo_() {}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>::ConstrainedQuadraticModel(
+        const ConstrainedQuadraticModel& other)
+        : objective(other.objective), constraints_(other.constraints_), varinfo_(other.varinfo_) {
+    objective.parent_ = this;
+    for (auto& c : constraints_) {
+        c.parent_ = this;
+    }
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>::ConstrainedQuadraticModel(
+        ConstrainedQuadraticModel&& other) noexcept
+        : ConstrainedQuadraticModel() {
+    swap(*this, other);
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>&
+ConstrainedQuadraticModel<bias_type, index_type>::operator=(ConstrainedQuadraticModel other) {
+    swap(*this, other);
+    return *this;
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>&
+ConstrainedQuadraticModel<bias_type, index_type>::operator=(
+        ConstrainedQuadraticModel&& other) noexcept {
+    swap(*this, other);
+    return *this;
+}
+
+template <class bias_type, class index_type>
+index_type ConstrainedQuadraticModel<bias_type, index_type>::add_constraint() {
+    constraints_.emplace_back(this);
+    return constraints_.size() - 1;
+}
+
+template <class bias_type, class index_type>
+template <class B, class I, class T>
+index_type ConstrainedQuadraticModel<bias_type, index_type>::add_constraint(
+        const abc::QuadraticModelBase<B, I>& lhs, Sense sense, bias_type rhs,
+        const std::vector<T>& mapping) {
+    // todo: move version
+    assert(mapping.size() == objective.num_variables());
+
+    add_constraint();
+    Constraint<bias_type, index_type>& constraint = constraints_.back();
+
+    for (size_type i = 0; i < lhs.num_variables(); ++i) {
+        assert(vartype(mapping[i]) == lhs.vartype(i));
+        assert(lower_bound(mapping[i]) == lhs.lower_bound(i));
+        assert(upper_bound(mapping[i]) == lhs.upper_bound(i));
+
+        constraint.add_linear(mapping[i], lhs.linear(i));
+    }
+
+    // quadratic biases
+    for (auto it = lhs.cbegin_quadratic(); it != lhs.cend_quadratic(); ++it) {
+        constraint.add_quadratic(mapping[it->u], mapping[it->v], it->bias);
+    }
+
+    // offset
+    constraint.add_offset(lhs.offset());
```

Comment:
Does clearing set the offset to zero ?   
Add offset or set offset ? Depending on the question above.

Reply:
This is a new one, so the offset starts at 0.